modules/collect/src/main/java/com/opengamma/collect/timeseries/DenseLocalDateDoubleTimeSeries.java

```diff
@@ -0,0 +1,984 @@
+    int endIndex = dateCalculation.calculatePosition(startDate, endExclusive);
+
+    return new DenseLocalDateDoubleTimeSeries(
+        startInclusive,
+        Arrays.copyOfRange(points, startIndex, endIndex),
+        dateCalculation);
+  }
+
+  @Override
+  public DenseLocalDateDoubleTimeSeries headSeries(int numPoints) {
+
+    ArgChecker.notNegative(numPoints, "numPoints");
+
+    if (numPoints == 0) {
+      return EMPTY_SERIES;
+    } else if (numPoints > size()) {
+      return this;
+    }
+
+    int endPosn = findHeadPoints(numPoints);
+
+    return new DenseLocalDateDoubleTimeSeries(startDate, Arrays.copyOf(points, endPosn), dateCalculation);
+  }
+
+  private int findHeadPoints(int required) {
+
+    // Take enough points that aren't NaN
+    // else we need the entire series
+    return IntStream.range(0, points.length)
+        .filter(i -> isValidPoint(points[i]))
+        .skip(required)
+        .findFirst()
+        .orElse(points.length);
+  }
+
+  @Override
+  public DenseLocalDateDoubleTimeSeries tailSeries(int numPoints) {
+    ArgChecker.notNegative(numPoints, "numPoints");
+
+    if (numPoints == 0) {
+      return EMPTY_SERIES;
+    } else if (numPoints > size()) {
+      return this;
+    }
+
+    int startPoint = findTailPoints(numPoints);
+
+    return new DenseLocalDateDoubleTimeSeries(
+        dateCalculation.calculateDateFromPosition(startDate, startPoint),
+        Arrays.copyOfRange(points, startPoint, points.length),
+        dateCalculation);
+  }
+
+  private int findTailPoints(int required) {
+
+    // As there is no way of constructing an IntStream from
+    // n to m where n > m, we go from -n to m and then
+    // take the additive inverse (sigh!)
+    return IntStream.rangeClosed(1 - points.length, 0)
+        .map(i -> -i)
+        .filter(i -> isValidPoint(points[i]))
+        .skip(required - 1)
+        .findFirst()
+        .orElse(0);
+  }
+
+  @Override
+  public Stream<LocalDateDoublePoint> stream() {
+    return validIndices()
+        .mapToObj(i -> LocalDateDoublePoint.of(dateCalculation.calculateDateFromPosition(startDate, i), points[i]));
+  }
+
+  @Override
+  public DoubleStream valueStream() {
+    return Arrays.stream(points).filter(this::isValidPoint);
+  }
+
+  @Override
+  public Stream<LocalDate> dateStream() {
+    return validIndices()
+        .mapToObj(i -> dateCalculation.calculateDateFromPosition(startDate, i));
+  }
+
+  private IntStream validIndices() {
+    return IntStream.range(0, points.length)
+        .filter(i -> isValidPoint(points[i]));
+  }
+
+  @Override
+  public DenseLocalDateDoubleTimeSeries filter(ObjDoublePredicate<LocalDate> predicate) {
+
+    DoubleStream filtered = IntStream.range(0, points.length)
+        .mapToDouble(i ->
+            isValidPoint(points[i]) &&
+                predicate.test(dateCalculation.calculateDateFromPosition(startDate, i), points[i]) ?
+            points[i] : Double.NaN);
```

Comment:
Wrapping makes this hard to understand. Perhaps a helper method and method reference?